integration/run_test.go

```diff
@@ -0,0 +1,356 @@
+    - REGISTRY_HTTP_ADDR={{.RegistryHost}}:443
+    volumes:
+    - {{.AuthDir}}:/auth:ro
+`, struct {
+		TargetStage     string
+		ServiceName     string
+		ImageContextDir string
+		RegistryHost    string
+		AuthDir         string
+	}{
+		TargetStage:     targetStage,
+		ServiceName:     serviceName,
+		ImageContextDir: pRoot,
+		RegistryHost:    regConfig.host,
+		AuthDir:         authDir,
+	}),
+		compose.WithBuildArgs(getBuildArgsFromEnv(t)...),
+		compose.WithStdio(testutil.TestingLogDest()))
+	if err != nil {
+		t.Fatalf("failed to prepare compose: %v", err)
+	}
+	defer c.Cleanup()
+	de, ok := c.Get(serviceName)
+	if !ok {
+		t.Fatalf("failed to get shell of service %v: %v", serviceName, err)
+	}
+	sh := shell.New(de, reporter)
+
+	// Initialize config files for containerd and snapshotter
+	additionalConfig := ""
+	if !isTestingBuiltinSnapshotter() {
+		additionalConfig = proxySnapshotterConfig
+	}
+	containerdConfigYaml := testutil.ApplyTextTemplate(t, `
+version = 2
+
+[plugins."io.containerd.snapshotter.v1.soci"]
+root_path = "/var/lib/soci-snapshotter-grpc/"
+
+[plugins."io.containerd.snapshotter.v1.soci".blob]
+check_always = true
+
+{{.AdditionalConfig}}
+`, struct {
+		RegistryHost     string
+		AdditionalConfig string
+	}{
+		RegistryHost:     regConfig.host,
+		AdditionalConfig: additionalConfig,
+	})
+
+	// Setup environment
+	if err := testutil.WriteFileContents(sh, defaultContainerdConfigPath, []byte(containerdConfigYaml), 0600); err != nil {
+		t.Fatalf("failed to write %v: %v", defaultContainerdConfigPath, err)
+	}
+	if err := testutil.WriteFileContents(sh, filepath.Join(caCertDir, "domain.crt"), crt, 0600); err != nil {
+		t.Fatalf("failed to write %v: %v", caCertDir, err)
+	}
+	sh.
+		X("update-ca-certificates").
+		Retry(100, "nerdctl", "login", "-u", regConfig.user, "-p", regConfig.pass, regConfig.host)
+
+	// imageNames := []string{nginxImage, drupalImage}
+	imageNames := []string{nginxImage}
+
+	rebootContainerd(t, sh, "", "")
+	for _, imageName := range imageNames {
+		// Mirror image
+		copyImage(sh, dockerhub(imageName), regConfig.mirror(imageName))
+		indexDigest := optimizeImage(sh, regConfig.mirror(imageName))
+
+		// Pull image
+		sh.X("ctr", "i", "pull", "--user", regConfig.creds(), regConfig.mirror(imageName).ref)
+
+		// Create SOCI metadata
+		sh.X("soci", "create", regConfig.mirror(imageName).ref)
+		sh.X("soci", "image", "rpull", "--user", regConfig.creds(), "--soci-index-digest", indexDigest, regConfig.mirror(imageName).ref)
+	}
+
+	// Run the containers
+	for _, imageName := range imageNames {
+		image := regConfig.mirror(imageName).ref
+		containerName := "test_" + regexp.MustCompile(`[^a-zA-Z0-9 ]+`).ReplaceAllString(imageName, "_")
+		sh.X("soci", "run", "-d", "--rm", "--snapshotter=soci", image, containerName)
+	}
+
+	time.Sleep(5 * time.Second) // FIXME find better solution for "connection refused" from `curl` below too soon
+
+	// Do something in each container
```

Comment:
If the `// Do something in each container` is the same in the two tests, maybe you can have a single function that does the testing logic (create images, create containers, check, etc). And probably for the two testcases, each can just pass in the function something like "which container uses which image".

Reply:
I hadn't originally intended both containers to be http servers. I'd probably want to move the "do something" out into a function that can be passed in, possibly on a per container basis?

Reply:
I've refactored into a single function for testing, the two Test functions are just one line each now.